File: 0111-minimum-depth-of-binary-tree/0111-minimum-depth-of-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def dfs(self, node: Optional[TreeNode], height: int) -> int:
        if not node.left and not node.right:
            return height
        ans = float('inf')
        if node.left:
            ans = min(ans, self.dfs(node.left, height + 1))
        if node.right:
            ans = min(ans, self.dfs(node.right, height + 1))
        return ans
    
    def minDepth(self, root: Optional[TreeNode]) -> int:
        if not root:
            return 0
        
        return self.dfs(root, 1)

File: 0111-minimum-depth-of-binary-tree/0111-minimum-depth-of-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int minDepth(TreeNode* root) {
        if (!root) return 0;
        
        queue<pair<TreeNode*, int>> queue;
        queue.push({root, 1});
        int ans = INT_MAX;
        
        while (!queue.empty()){
            auto [node, height] = queue.front();
            queue.pop();
            
            if (!node -> left && !node -> right) ans = min(ans, height);
            
            if (node -> left) queue.push({node -> left, height + 1});
            if (node -> right) queue.push({node -> right, height + 1});
        }
        
        return ans;
    }
};